core/controller/src/main/scala/whisk/core/controller/Actions.scala

```diff
@@ -490,21 +495,35 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
      * If this mechanism fails to produce an answer quickly, the future will fail and we back off into using
      * a database operation to obtain the canonical answer.
      */
-    private def pollLocalForResult(
-        docid: DocInfo,
+    private def waitForActivationResponse(
+        namespace: Namespace,
         activationId: ActivationId,
-        promise: Promise[Option[WhiskActivation]])(
-            implicit transid: TransactionId): Unit = {
-        queryActivationResponse(activationId, transid) map {
-            activation => promise.trySuccess { Some(activation) }
-        } onFailure {
+        timeout: FiniteDuration)(
+            implicit transid: TransactionId): (ActivationId, Future[WhiskActivation]) = {
+
+        val (activationIdToSearch, activeCallback) = queryActivationResponse(activationId, transid)
+
+        val promise = Promise[WhiskActivation]
```

Comment:
Can we get rid of this promise aswell? It doesn't strike me as necessary to have the promise passed to `pollDbForResult`.

`recoverWith` might be helpful